### PR TITLE
Add coverage testing.

### PR DESCRIPTION
**Pull request by LukeShu**
This patchset adds coverage testing.  After running `make check`, the master coverfile is `test/cov.cov`, and an HTML report is generated from it at `test/cov.html`.

This "vendors" in `util/` two new utilities: `write-ifchanged`, which I've been copying from project to project for several years, and [`gocovcat`](https://git.lukeshu.com/go/cmd/gocovcat/). Though I've since created a separate repository for gocovcat, it actually originated in holo and I git-filter-branch'ed it out.

The most invasive part of this change is that it switches over to use the normal Go workspace structure.  This means that output binaries go in `bin/` instead of `build/`.  Because `build/` no longer exists, I've moved the output location for manpages to `man/`.

### CODE REVIEW

**Comment by LukeShu** [comment]
BTW, Adding a coveralls.io badge to the README would be easy, but would require you to create an account and add the repo there.

**Comment by majewsky** [comment]
Wow. I will have to take my time to dig through this. Do you have a reference link for the "normal Go workspace structure" for me? I've not seen this convention before.

**Comment by LukeShu** [comment]
Perhaps you just haven't heard the term "workspace" before (it seems to be less common); it refers to a directory pointed to by `$GOPATH`.

Workspace reference: https://golang.org/doc/code.html#Workspaces
`go` command reference saying that relative import paths are not allowed in workspaces: https://golang.org/cmd/go/#hdr-Relative_import_paths

File: Makefile [hunk 1]
```diff
@@ -1,56 +1,72 @@
...
+
+man:
+	@mkdir $@
 
 # manpages are generated using pod2man (which comes with Perl and therefore
 # should be readily available on almost every Unix system)
-build/man/%: doc/%.pod
+man/%: doc/%.pod | man
 	pod2man --name="$(shell echo $* | cut -d. -f1)" --section=$(shell echo $* | cut -d. -f2) \
 		--center="Configuration Management" --release="Holo $(VERSION)" \
```
Comment (majewsky) [inline Makefile]: The `$(VERSION)` interpolation is broken after you changed the usage of `util/find_version.sh`.

File: src/holo-files/main_test.go [hunk 2]
```diff
@@ -0,0 +1 @@
+../main.go.test
```
Comment (majewsky) [inline src/holo-files/main_test.go]: What is the purpose of this test (and of `src/holo/main_test.go`)?

File: src/main.go.test [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	}
+
+	flag.CommandLine.Parse(flags)
+	_ = m.Run()
+	os.Exit(exit)
+}
+
+func TestSystem(t *testing.T) {
+	exit = Main()
+	_ = os.Stdout.Close()
```
Comment (majewsky) [inline src/main.go.test]: You had some additional comments about `Main()` and `os.Stdout.Close` in the message for this commit. Can you add these into the code instead?

**Review by majewsky — COMMENTED** [review]
Please find below some nitpicks. I will add another comment about the big picture, which doesn't fit in the small textbox that Github offers me right now. :)

**Comment by majewsky** [comment]
Big picture issue no. 1: I'm not satisfied with the workspace structure as it is established in this branch. There are a lot of properties that I like, esp. "no relative imports". That was just the younger me not knowing how to do things properly.

I've seen a lot of other people's Go repos in the previous months, and the general pattern for how they organize their code [1] uses directories `<repo-root>/cmd/<binary>` for each binary, and groups the shared code into `<repo-root>/pkg`. Imports are always done as `import "github.com/<org>/<repo>/pkg/<module>"`.

Compared to that established pattern, a line like `import "holo-files/common"` looks really surprising. Out of context, I would wonder why holo-files is part of the stdlib now. Module names that don't comply with the layout of the user's regular GOPATH also shut the door on ever trying to import these modules somewhere else.

I will try to do a competing PR with the repo layout that is floating in my head right now, so we can discuss the merits of both versions against each other.

[1] As an example, see https://github.com/kubernetes/helm.

Big picture issue no. 2: There's few documentation on all the clever contraptions that you designed to make the coverage testing work. I'm willing to merge the PR without docs, though. It will be a nice exercise to write the documentation myself and have you review if I really understood everything. :)

**Comment by LukeShu** [comment]
A previous iteration of my coverage-testing branch actually did move everything into `holocm.org/cmd` (and `holocm.org/lib`), but I ended up rebasing that out to minimize the size of the PR to the minimum necessary to get coverage testing working.

**Comment by LukeShu** [comment]
I think seeing `pkg` in the root of the project is confusing; after working with Go a bunch, I expect `pkg` to be the directory with compiled library files.

File: src/holo-files/main_test.go (hunk 2 above)
Comment (LukeShu) [reply src/holo-files/main_test.go]: The binaries produced by `go test` don't actually call `main()`; instead, you need to define a test that calls your main function.

**Comment by LukeShu** [comment]
I've definitely got writing a blog post about the coverage testing on my TODO list :)

My implementation was based heavily on [this article from Elastic](https://www.elastic.co/blog/code-coverage-for-your-golang-system-tests), but [this](https://blog.cloudflare.com/go-coverage-with-external-tests/) article would also have been very helpful if it had existed at the time.

**Comment by LukeShu** [comment]
Oh, if you don't like having the "vendored" `gocovcat` in `util/`, we could add an external dependency on it, or on github.com/wadey/gocovmerge (which as I understand it does the same thing).  Perhaps via a git submodule, if you don't think those are too awful.

**Comment by majewsky** [comment]
I'm fine with the vendored gocovcat, following the precedent of `util/find_version.sh` being copied into most other repos in this org.

Re git submodules, I've had quite the negative experience them, esp. because they are not considered by Github when it creates release tarballs. See https://github.com/holocm/holo-attic/blob/master/util/make-release-tarball.sh for the hairs that I pulled out over this.

**Comment by majewsky** [comment]
Now that #28 is merged, please rebase.

**Comment by LukeShu** [comment]
In rebasing this, I learned that the `go` command doesn't follow symlinks when evaluating `...` wildcards (see `go help packages`, though it doesn't mention links); which I used to list the packages to coverage test for.

So it will take me a bit to work around `$GOPATH/src/github.com/holocm/holo` being a symlink.

**Comment by LukeShu** [comment]
The command to build the coverage-testing binaries is now pretty gross.  It uses `$(go list)` with some arguments to enumerate the packages that it is using.

**Comment by LukeShu** [comment]
Ping?

**Comment by majewsky** [comment]
Oh yeah, sorry for dropping the ball.

**Comment by majewsky** [comment]
I've also set up coveralls.io and added the badge to the README.

**Pull request merged**